**Import Libraries:**

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

**Upload the file**

In [ ]:
df_movies=pd.read_csv("/content/movie_metadata.csv")
org_data=df_movies

In [ ]:
df_movies.head(2)   # Display top 2 records

color   director_name  num_critic_for_reviews  duration  \
0  Color   James Cameron                   723.0     178.0   
1  Color  Gore Verbinski                   302.0     169.0   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                      0.0                   855.0  Joel David Moore   
1                    563.0                  1000.0     Orlando Bloom   

   actor_1_facebook_likes        gross                           genres  ...  \
0                  1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi  ...   
1                 40000.0  309404152.0         Action|Adventure|Fantasy  ...   

  num_user_for_reviews language  country  content_rating       budget  \
0               3054.0  English      USA           PG-13  237000000.0   
1               1238.0  English      USA           PG-13  300000000.0   

   title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0      2009.0                  936.0        7.9          1.78   
1      2007.0                 5000.0        7.1          2.35   

  movie_facebook_likes  
0                33000  
1                    0  

[2 rows x 28 columns]

In [ ]:
df_movies.shape   # To show shape of the dataframe

(5043, 28)

In [ ]:
df_movies.info()   # Display information abt dataset

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5043 entries, 0 to 5042
Data columns (total 28 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   color                      5024 non-null   object 
 1   director_name              4939 non-null   object 
 2   num_critic_for_reviews     4993 non-null   float64
 3   duration                   5028 non-null   float64
 4   director_facebook_likes    4939 non-null   float64
 5   actor_3_facebook_likes     5020 non-null   float64
 6   actor_2_name               5030 non-null   object 
 7   actor_1_facebook_likes     5036 non-null   float64
 8   gross                      4159 non-null   float64
 9   genres                     5043 non-null   object 
 10  actor_1_name               5036 non-null   object 
 11  movie_title                5043 non-null   object 
 12  num_voted_users            5043 non-null   int64  
 13  cast_total_facebook_likes  5043 non-null   int64

**Cleaning Data**

In [ ]:
df_movies.isnull().sum().sort_values(ascending=False)  # Find the missing values column wise

gross                        884
budget                       492
aspect_ratio                 329
content_rating               303
plot_keywords                153
title_year                   108
director_name                104
director_facebook_likes      104
num_critic_for_reviews        50
actor_3_name                  23
actor_3_facebook_likes        23
num_user_for_reviews          21
color                         19
duration                      15
facenumber_in_poster          13
actor_2_name                  13
actor_2_facebook_likes        13
language                      12
actor_1_name                   7
actor_1_facebook_likes         7
country                        5
cast_total_facebook_likes      0
num_voted_users                0
movie_title                    0
movie_imdb_link                0
genres                         0
imdb_score                     0
movie_facebook_likes           0
dtype: int64

In [ ]:
df_movies.isnull().sum(axis=1).sort_values(ascending=False)  # Find the missing values row wise

279     15
4       14
4945    11
2241    11
2342    10
        ..
1703     0
1702     0
1701     0
1700     0
5042     0
Length: 5043, dtype: int64

In [ ]:
df_movies.isnull().sum().sort_values(ascending=False)/len(df_movies)*100   # find column wise null percentage

gross                        17.529248
budget                        9.756098
aspect_ratio                  6.523895
content_rating                6.008328
plot_keywords                 3.033908
title_year                    2.141582
director_name                 2.062265
director_facebook_likes       2.062265
num_critic_for_reviews        0.991473
actor_3_name                  0.456078
actor_3_facebook_likes        0.456078
num_user_for_reviews          0.416419
color                         0.376760
duration                      0.297442
facenumber_in_poster          0.257783
actor_2_name                  0.257783
actor_2_facebook_likes        0.257783
language                      0.237954
actor_1_name                  0.138806
actor_1_facebook_likes        0.138806
country                       0.099147
cast_total_facebook_likes     0.000000
num_voted_users               0.000000
movie_title                   0.000000
movie_imdb_link               0.000000
genres                   

### **Drop the unnecessasry columns**

In [ ]:
df_movies=df_movies.drop([
    'color','director_facebook_likes','actor_3_facebook_likes','actor_1_facebook_likes','actor_2_facebook_likes',
    'cast_total_facebook_likes','facenumber_in_poster','content_rating','plot_keywords',
    'actor_2_name','actor_3_name','duration','aspect_ratio','movie_imdb_link','country'
],axis=1)

In [ ]:
df_movies.shape

(5043, 13)

### **Drop unnecessary rows**

In [ ]:
# Drop unnecessary rows (rows which have more null values)
round(df_movies.isnull().sum().sort_values(ascending=False)/len(df_movies)*100,2)

gross                     17.53
budget                     9.76
title_year                 2.14
director_name              2.06
num_critic_for_reviews     0.99
num_user_for_reviews       0.42
language                   0.24
actor_1_name               0.14
genres                     0.00
movie_title                0.00
num_voted_users            0.00
imdb_score                 0.00
movie_facebook_likes       0.00
dtype: float64

In [ ]:
# Clean the data of columns gross and budget
df_movies=df_movies[df_movies['gross'].notnull()]
df_movies=df_movies[df_movies['budget'].notnull()]


In [ ]:
# Again check the data of rows (rows which have more null values)
round(df_movies.isnull().sum().sort_values(ascending=False)/len(df_movies)*100,2)

actor_1_name              0.08
language                  0.08
num_critic_for_reviews    0.03
director_name             0.00
gross                     0.00
genres                    0.00
movie_title               0.00
num_voted_users           0.00
num_user_for_reviews      0.00
budget                    0.00
title_year                0.00
imdb_score                0.00
movie_facebook_likes      0.00
dtype: float64

In [ ]:
df_movies.isnull().sum(axis=1).sort_values(ascending=False)   # Check the null values of rows

4502    1
4720    1
4110    1
4837    1
4711    1
       ..
1375    0
1376    0
1377    0
1378    0
5042    0
Length: 3891, dtype: int64

In [ ]:
# Check the null values of rows which have more than 5
(df_movies.isnull().sum(axis=1).sort_values(ascending=False)>5).sum()

0

In [ ]:
df_movies=df_movies[df_movies.isnull().sum(axis=1).sort_values(ascending=False)<=5]

<ipython-input-17-eb72336f3e13>:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df_movies=df_movies[df_movies.isnull().sum(axis=1).sort_values(ascending=False)<=5]


### **Fill NAN Values**

In [ ]:
round(df_movies.isnull().sum().sort_values(ascending=False)/len(df_movies)*100,2)

actor_1_name              0.08
language                  0.08
num_critic_for_reviews    0.03
director_name             0.00
gross                     0.00
genres                    0.00
movie_title               0.00
num_voted_users           0.00
num_user_for_reviews      0.00
budget                    0.00
title_year                0.00
imdb_score                0.00
movie_facebook_likes      0.00
dtype: float64

In [ ]:
df_movies['language'].value_counts().nlargest(5)

English     3707
French        37
Spanish       26
Mandarin      15
German        13
Name: language, dtype: int64

In [ ]:
df_movies.language.describe()

count        3888
unique         38
top       English
freq         3707
Name: language, dtype: object

**Observations:** Here total 3888 movies in that 3707 movies are English movies remaining movies are related to different(37) languages.

In [ ]:
df_movies.language=df_movies.language.fillna('English')

In [ ]:
round(df_movies.isnull().sum().sort_values(ascending=False)/len(df_movies)*100,2)

actor_1_name              0.08
num_critic_for_reviews    0.03
director_name             0.00
gross                     0.00
genres                    0.00
movie_title               0.00
num_voted_users           0.00
num_user_for_reviews      0.00
language                  0.00
budget                    0.00
title_year                0.00
imdb_score                0.00
movie_facebook_likes      0.00
dtype: float64

###**Check the number of retained rows**

You might notice that two of the columns num_critic_for_reviews and actor_1_name have small percentage of NaN valuesleft. You can let these columns as it is for now.Check the number and percentage of the rows retained after completing all the tasks above.

In [ ]:
len(df_movies)

3891

In [ ]:
# Checking the no.of  retained rows here ( retained meaning-continue to hold )
round(len(df_movies)/len(org_data)*100,2)

77.16

**Observarions:** You might have noticed that we still have around 77% of rows.

## **Data Analysis**

**Change the unit of columns**



In [ ]:
df_movies['budget']/1000000

0       237.0000
1       300.0000
2       245.0000
3       250.0000
5       263.7000
          ...   
5033      0.0070
5034      0.0070
5035      0.0070
5037      0.0090
5042      0.0011
Name: budget, Length: 3891, dtype: float64

In [ ]:
# Convert the unit of the budget and gross columns from $ to million $
df_movies['budget']=df_movies['budget']/1000000
df_movies['gross']=df_movies['gross']/1000000

**Find the movies with highest profit**

Create a new column called profit which contains differece of the two columns gross and budget

In [ ]:
df_movies['profit']=df_movies['gross']-df_movies['budget']
df_movies

director_name  num_critic_for_reviews       gross  \
0         James Cameron                   723.0  760.505847   
1        Gore Verbinski                   302.0  309.404152   
2            Sam Mendes                   602.0  200.074175   
3     Christopher Nolan                   813.0  448.130642   
5        Andrew Stanton                   462.0   73.058679   
...                 ...                     ...         ...   
5033      Shane Carruth                   143.0    0.424760   
5034   Neill Dela Llana                    35.0    0.070071   
5035   Robert Rodriguez                    56.0    2.040920   
5037       Edward Burns                    14.0    0.004584   
5042           Jon Gunn                    43.0    0.085222   

                                   genres     actor_1_name  \
0         Action|Adventure|Fantasy|Sci-Fi      CCH Pounder   
1                Action|Adventure|Fantasy      Johnny Depp   
2               Action|Adventure|Thriller  Christoph Waltz   
3                         Action|Thriller        Tom Hardy   
5                 Action|Adventure|Sci-Fi     Daryl Sabara   
...                                   ...              ...   
5033                Drama|Sci-Fi|Thriller    Shane Carruth   
5034                             Thriller      Ian Gamazon   
5035  Action|Crime|Drama|Romance|Thriller  Carlos Gallardo   
5037                         Comedy|Drama      Kerry Bishé   
5042                          Documentary      John August   

                                    movie_title  num_voted_users  \
0                                       Avatar            886204   
1     Pirates of the Caribbean: At World's End            471220   
2                                      Spectre            275868   
3                        The Dark Knight Rises           1144337   
5                                  John Carter            212204   
...                                         ...              ...   
5033                                    Primer             72639   
5034                                    Cavite               589   
5035                               El Mariachi             52055   
5037                                 Newlyweds              1338   
5042                         My Date with Drew              4285   

      num_user_for_reviews language    budget  title_year  imdb_score  \
0                   3054.0  English  237.0000      2009.0         7.9   
1                   1238.0  English  300.0000      2007.0         7.1   
2                    994.0  English  245.0000      2015.0         6.8   
3                   2701.0  English  250.0000      2012.0         8.5   
5                    738.0  English  263.7000      2012.0         6.6   
...                    ...      ...       ...         ...         ...   
5033                 371.0  English    0.0070      2004.0         7.0   
5034                  35.0  English    0.0070      2005.0         6.3   
5035                 130.0  Spanish    0.0070      1992.0         6.9   
5037                  14.0  English    0.0090      2011.0         6.4   
5042                  84.0  English    0.0011      2004.0         6.6   

      movie_facebook_likes      profit  
0                    33000  523.505847  
1                        0    9.404152  
2                    85000  -44.925825  
3                   164000  198.130642  
5                    24000 -190.641321  
...                    ...         ...  
5033                 19000    0.417760  
5034                    74    0.063071  
5035                     0    2.033920  
5037                   413   -0.004416  
5042                   456    0.084122  

[3891 rows x 14 columns]

**Observations:** Here we see a new column profit added at end ot the data set.In profit some values shown in negative means the movies which got loss.

In [ ]:
# Set the dataframe using the profit column as reference.
df_movies.sort_values(by='profit',ascending=False)

director_name  num_critic_for_reviews       gross  \
0        James Cameron                   723.0  760.505847   
29     Colin Trevorrow                   644.0  652.177271   
26       James Cameron                   315.0  658.672302   
3024      George Lucas                   282.0  460.935665   
3080  Steven Spielberg                   215.0  434.949459   
...                ...                     ...         ...   
2334   Katsuhiro Ôtomo                   105.0    0.410388   
2323    Hayao Miyazaki                   174.0    2.298191   
3005      Lajos Koltai                    73.0    0.195888   
3859    Chan-wook Park                   202.0    0.211667   
2988      Joon-ho Bong                   363.0    2.201412   

                                                 genres         actor_1_name  \
0                       Action|Adventure|Fantasy|Sci-Fi          CCH Pounder   
29                     Action|Adventure|Sci-Fi|Thriller  Bryce Dallas Howard   
26                                        Drama|Romance    Leonardo DiCaprio   
3024                    Action|Adventure|Fantasy|Sci-Fi        Harrison Ford   
3080                                      Family|Sci-Fi         Henry Thomas   
...                                                 ...                  ...   
2334  Action|Adventure|Animation|Family|Sci-Fi|Thriller     William Hootkins   
2323                        Adventure|Animation|Fantasy        Minnie Driver   
3005                                  Drama|Romance|War         Marcell Nagy   
3859                                        Crime|Drama         Min-sik Choi   
2988                         Comedy|Drama|Horror|Sci-Fi            Doona Bae   

                              movie_title  num_voted_users  \
0                                 Avatar            886204   
29                        Jurassic World            418214   
26                               Titanic            793059   
3024  Star Wars: Episode IV - A New Hope            911097   
3080          E.T. the Extra-Terrestrial            281842   
...                                   ...              ...   
2334                            Steamboy             13727   
2323                   Princess Mononoke            221552   
3005                            Fateless              5603   
3859                      Lady Vengeance             53508   
2988                            The Host             68883   

      num_user_for_reviews   language        budget  title_year  imdb_score  \
0                   3054.0    English    237.000000      2009.0         7.9   
29                  1290.0    English    150.000000      2015.0         7.0   
26                  2528.0    English    200.000000      1997.0         7.7   
3024                1470.0    English     11.000000      1977.0         8.7   
3080                 515.0    English     10.500000      1982.0         7.9   
...                    ...        ...           ...         ...         ...   
2334                  79.0   Japanese   2127.519898      2004.0         6.9   
2323                 570.0   Japanese   2400.000000      1997.0         8.4   
3005                  45.0  Hungarian   2500.000000      2005.0         7.1   
3859                 131.0     Korean   4200.000000      2005.0         7.7   
2988                 279.0     Korean  12215.500000      2006.0         7.0   

      movie_facebook_likes        profit  
0                    33000    523.505847  
29                  150000    502.177271  
26                   26000    458.672302  
3024                 33000    449.935665  
3080                 34000    424.449459  
...                    ...           ...  
2334                   973  -2127.109510  
2323                 11000  -2397.701809  
3005                   607  -2499.804112  
3859                  4000  -4199.788333  
2988                  7000 -12213.298588  

[3891 rows x 14 columns]

In [ ]:
#Extract the top ten profiting movies in descending order and store them in a new dataframe-top10
top10 =df_movies.sort_values(by='profit',ascending=False).head(10)
top10

director_name  num_critic_for_reviews       gross  \
0         James Cameron                   723.0  760.505847   
29      Colin Trevorrow                   644.0  652.177271   
26        James Cameron                   315.0  658.672302   
3024       George Lucas                   282.0  460.935665   
3080   Steven Spielberg                   215.0  434.949459   
794         Joss Whedon                   703.0  623.279547   
17          Joss Whedon                   703.0  623.279547   
509        Roger Allers                   186.0  422.783777   
240        George Lucas                   320.0  474.544677   
66    Christopher Nolan                   645.0  533.316061   

                                        genres         actor_1_name  \
0              Action|Adventure|Fantasy|Sci-Fi          CCH Pounder   
29            Action|Adventure|Sci-Fi|Thriller  Bryce Dallas Howard   
26                               Drama|Romance    Leonardo DiCaprio   
3024           Action|Adventure|Fantasy|Sci-Fi        Harrison Ford   
3080                             Family|Sci-Fi         Henry Thomas   
794                    Action|Adventure|Sci-Fi      Chris Hemsworth   
17                     Action|Adventure|Sci-Fi      Chris Hemsworth   
509   Adventure|Animation|Drama|Family|Musical    Matthew Broderick   
240            Action|Adventure|Fantasy|Sci-Fi      Natalie Portman   
66                 Action|Crime|Drama|Thriller       Christian Bale   

                                     movie_title  num_voted_users  \
0                                        Avatar            886204   
29                               Jurassic World            418214   
26                                      Titanic            793059   
3024         Star Wars: Episode IV - A New Hope            911097   
3080                 E.T. the Extra-Terrestrial            281842   
794                                The Avengers            995415   
17                                 The Avengers            995415   
509                               The Lion King            644348   
240   Star Wars: Episode I - The Phantom Menace            534658   
66                              The Dark Knight           1676169   

      num_user_for_reviews language  budget  title_year  imdb_score  \
0                   3054.0  English   237.0      2009.0         7.9   
29                  1290.0  English   150.0      2015.0         7.0   
26                  2528.0  English   200.0      1997.0         7.7   
3024                1470.0  English    11.0      1977.0         8.7   
3080                 515.0  English    10.5      1982.0         7.9   
794                 1722.0  English   220.0      2012.0         8.1   
17                  1722.0  English   220.0      2012.0         8.1   
509                  656.0  English    45.0      1994.0         8.5   
240                 3597.0  English   115.0      1999.0         6.5   
66                  4667.0  English   185.0      2008.0         9.0   

      movie_facebook_likes      profit  
0                    33000  523.505847  
29                  150000  502.177271  
26                   26000  458.672302  
3024                 33000  449.935665  
3080                 34000  424.449459  
794                 123000  403.279547  
17                  123000  403.279547  
509                  17000  377.783777  
240                  13000  359.544677  
66                   37000  348.316061

**Observations:** After you found out top 10 movies, you might have noticed duplicate values at row 17 and 794. It seems the dataframe has duplicate values as well.Drop the duplicate values.

###Drop Duplicate Values

In [ ]:
df_movies.drop_duplicates(keep='first',inplace=True)

In [ ]:
df_movies.shape

(3856, 14)

##Find IMDB Top 250

Create the new dataframe IMDB_Top_250 and store the top 250 movies with highest IMDB rating(corresponding the column imdb_score). Also make sure that all of these movies num_voted_users greater than 25000.Also add a rank column containing the values 1 to 250 indicating the ranks of the corresponding films.

In [ ]:
IMDB_Top_250 =df_movies[df_movies['num_voted_users']>25000].sort_values(by='imdb_score',ascending=False).head(250)
IMDB_Top_250

director_name  num_critic_for_reviews       gross  \
1937        Frank Darabont                   199.0   28.341469   
3466  Francis Ford Coppola                   208.0  134.821952   
2837  Francis Ford Coppola                   149.0   57.300000   
66       Christopher Nolan                   645.0  533.316061   
4498          Sergio Leone                   181.0    6.100000   
...                    ...                     ...         ...   
4931           John Carney                   232.0    9.437933   
2605               Ang Lee                   287.0  128.067808   
3029      David O. Russell                   410.0   93.571803   
2177            Tim Burton                   111.0   56.362352   
2487          George Cukor                    82.0   72.000000   

                            genres    actor_1_name  \
1937                   Crime|Drama  Morgan Freeman   
3466                   Crime|Drama       Al Pacino   
2837                   Crime|Drama  Robert De Niro   
66     Action|Crime|Drama|Thriller  Christian Bale   
4498                       Western  Clint Eastwood   
...                            ...             ...   
4931           Drama|Music|Romance    Glen Hansard   
2605          Action|Drama|Romance      Chen Chang   
3029         Biography|Drama|Sport  Christian Bale   
2177               Fantasy|Romance     Johnny Depp   
2487  Drama|Family|Musical|Romance    Jeremy Brett   

                          movie_title  num_voted_users  num_user_for_reviews  \
1937        The Shawshank Redemption           1689764                4144.0   
3466                   The Godfather           1155770                2238.0   
2837          The Godfather: Part II            790926                 650.0   
66                   The Dark Knight           1676169                4667.0   
4498  The Good, the Bad and the Ugly            503509                 780.0   
...                               ...              ...                   ...   
4931                            Once             90827                 329.0   
2605  Crouching Tiger, Hidden Dragon            217740                1641.0   
3029                     The Fighter            275869                 389.0   
2177             Edward Scissorhands            357581                 588.0   
2487                    My Fair Lady             66959                 258.0   

      language  budget  title_year  imdb_score  movie_facebook_likes  \
1937   English   25.00      1994.0         9.3                108000   
3466   English    6.00      1972.0         9.2                 43000   
2837   English   13.00      1974.0         9.0                 14000   
66     English  185.00      2008.0         9.0                 37000   
4498   Italian    1.20      1966.0         8.9                 20000   
...        ...     ...         ...         ...                   ...   
4931   English    0.18      2007.0         7.9                 26000   
2605  Mandarin   15.00      2000.0         7.9                     0   
3029   English   25.00      2010.0         7.9                 36000   
2177   English   20.00      1990.0         7.9                 16000   
2487   English   17.00      1964.0         7.9                     0   

          profit  
1937    3.341469  
3466  128.821952  
2837   44.300000  
66    348.316061  
4498    4.900000  
...          ...  
4931    9.257933  
2605  113.067808  
3029   68.571803  
2177   36.362352  
2487   55.000000  

[250 rows x 14 columns]

In [ ]:
# Add a new column called Rank
IMDB_Top_250['Rank']=IMDB_Top_250['imdb_score'].rank(method='first',ascending=False)
IMDB_Top_250

director_name  num_critic_for_reviews       gross  \
1937        Frank Darabont                   199.0   28.341469   
3466  Francis Ford Coppola                   208.0  134.821952   
2837  Francis Ford Coppola                   149.0   57.300000   
66       Christopher Nolan                   645.0  533.316061   
4498          Sergio Leone                   181.0    6.100000   
...                    ...                     ...         ...   
4931           John Carney                   232.0    9.437933   
2605               Ang Lee                   287.0  128.067808   
3029      David O. Russell                   410.0   93.571803   
2177            Tim Burton                   111.0   56.362352   
2487          George Cukor                    82.0   72.000000   

                            genres    actor_1_name  \
1937                   Crime|Drama  Morgan Freeman   
3466                   Crime|Drama       Al Pacino   
2837                   Crime|Drama  Robert De Niro   
66     Action|Crime|Drama|Thriller  Christian Bale   
4498                       Western  Clint Eastwood   
...                            ...             ...   
4931           Drama|Music|Romance    Glen Hansard   
2605          Action|Drama|Romance      Chen Chang   
3029         Biography|Drama|Sport  Christian Bale   
2177               Fantasy|Romance     Johnny Depp   
2487  Drama|Family|Musical|Romance    Jeremy Brett   

                          movie_title  num_voted_users  num_user_for_reviews  \
1937        The Shawshank Redemption           1689764                4144.0   
3466                   The Godfather           1155770                2238.0   
2837          The Godfather: Part II            790926                 650.0   
66                   The Dark Knight           1676169                4667.0   
4498  The Good, the Bad and the Ugly            503509                 780.0   
...                               ...              ...                   ...   
4931                            Once             90827                 329.0   
2605  Crouching Tiger, Hidden Dragon            217740                1641.0   
3029                     The Fighter            275869                 389.0   
2177             Edward Scissorhands            357581                 588.0   
2487                    My Fair Lady             66959                 258.0   

      language  budget  title_year  imdb_score  movie_facebook_likes  \
1937   English   25.00      1994.0         9.3                108000   
3466   English    6.00      1972.0         9.2                 43000   
2837   English   13.00      1974.0         9.0                 14000   
66     English  185.00      2008.0         9.0                 37000   
4498   Italian    1.20      1966.0         8.9                 20000   
...        ...     ...         ...         ...                   ...   
4931   English    0.18      2007.0         7.9                 26000   
2605  Mandarin   15.00      2000.0         7.9                     0   
3029   English   25.00      2010.0         7.9                 36000   
2177   English   20.00      1990.0         7.9                 16000   
2487   English   17.00      1964.0         7.9                     0   

          profit   Rank  
1937    3.341469    1.0  
3466  128.821952    2.0  
2837   44.300000    3.0  
66    348.316061    4.0  
4498    4.900000    5.0  
...          ...    ...  
4931    9.257933  246.0  
2605  113.067808  247.0  
3029   68.571803  248.0  
2177   36.362352  249.0  
2487   55.000000  250.0  

[250 rows x 15 columns]

Extract all the movies in the IMDB_Top_250 dataframe which are not in the English language and store them in a new dataframe named **Top_Foreign_lang_film**

In [ ]:
Top_Foreign_lang_film=IMDB_Top_250[IMDB_Top_250['language']!='English']
Top_Foreign_lang_film

director_name  num_critic_for_reviews       gross  \
4498                      Sergio Leone                   181.0    6.100000   
4747                    Akira Kurosawa                   153.0    0.269061   
4029                Fernando Meirelles                   214.0    7.563397   
2373                    Hayao Miyazaki                   246.0   10.049886   
4259  Florian Henckel von Donnersmarck                   215.0   11.284657   
4921                      Majid Majidi                    46.0    0.925402   
2323                    Hayao Miyazaki                   174.0    2.298191   
2970                 Wolfgang Petersen                    96.0   11.433134   
4105                    Chan-wook Park                   305.0    2.181290   
4659                    Asghar Farhadi                   354.0    7.098492   
1329                    S.S. Rajamouli                    44.0    6.498000   
1298                Jean-Pierre Jeunet                   242.0   33.201661   
2734                        Fritz Lang                   260.0    0.026435   
4033                 Thomas Vinterberg                   349.0    0.610968   
2829               Oliver Hirschbiegel                   192.0    5.501940   
2551                Guillermo del Toro                   406.0   37.623143   
4000              Juan José Campanella                   262.0   20.167424   
3550                  Denis Villeneuve                   226.0    6.857096   
2047                    Hayao Miyazaki                   212.0    4.710455   
2830                Alejandro Amenábar                   157.0    2.086345   
2914                       Je-kyu Kang                    86.0    1.110186   
4461                 Thomas Vinterberg                    98.0    1.647780   
3553                      José Padilha                   142.0    0.008060   
3423                   Katsuhiro Ôtomo                   150.0    0.439162   
4267             Alejandro G. Iñárritu                   157.0    5.383834   
3456                 Vincent Paronnaud                   242.0    4.443403   
3344                       Karan Johar                   210.0    4.018695   
4144                     Walter Salles                    71.0    5.595428   
4284                        Ari Folman                   231.0    2.283276   
4897                      Sergio Leone                   122.0    3.500000   
1171                       Yimou Zhang                   283.0    0.084961   
2863                    Clint Eastwood                   251.0   13.753931   
3264                    Michael Haneke                   447.0    0.225377   
3510                       Yash Chopra                    29.0    2.921738   
3677              Christophe Barratier                   112.0    3.629758   
4415                  Fabián Bielinsky                    94.0    1.221261   
4640                   Cristian Mungiu                   233.0    1.185783   
2605                           Ang Lee                   287.0  128.067808   

                                                 genres  \
4498                                            Western   
4747                             Action|Adventure|Drama   
4029                                        Crime|Drama   
2373                 Adventure|Animation|Family|Fantasy   
4259                                     Drama|Thriller   
4921                                       Drama|Family   
2323                        Adventure|Animation|Fantasy   
2970                       Adventure|Drama|Thriller|War   
4105                             Drama|Mystery|Thriller   
4659                                      Drama|Mystery   
1329                 Action|Adventure|Drama|Fantasy|War   
1298                                     Comedy|Romance   
2734                                       Drama|Sci-Fi   
4033                                              Drama   
2829                        Biography|Drama|History|War   
2551                                  

### **Find the best directors**

In [ ]:
# Extracting top 10 directors
top10_directors = df_movies.groupby('director_name').imdb_score.mean().sort_values(ascending=False).head(10)
top10_directors

director_name
Charles Chaplin          8.600000
Tony Kaye                8.600000
Alfred Hitchcock         8.500000
Ron Fricke               8.500000
Damien Chazelle          8.500000
Majid Majidi             8.500000
Sergio Leone             8.433333
Christopher Nolan        8.425000
S.S. Rajamouli           8.400000
Marius A. Markevicius    8.400000
Name: imdb_score, dtype: float64

### Find popular Genres

Extract first two genres from the genres column and store them in two new columns genre_1 and genre_2 . Some of the movies might have only one genre. In such cases, extract the single genre into both the columns.

In [ ]:
Temp_genre=df_movies.genres.str.split('|',expand=True).iloc[:,0:2]
Temp_genre.columns=['genre_1','genre_2']   # Set label to columns
Temp_genre

genre_1    genre_2
0          Action  Adventure
1          Action  Adventure
2          Action  Adventure
3          Action   Thriller
5          Action  Adventure
...           ...        ...
5033        Drama     Sci-Fi
5034     Thriller       None
5035       Action      Crime
5037       Comedy      Drama
5042  Documentary       None

[3856 rows x 2 columns]

**Observations:** Here genre_2 column have some null values, so fill those values with genre_1 values.

In [ ]:
Temp_genre['genre_2'].fillna(Temp_genre['genre_1'],inplace=True)
Temp_genre

genre_1      genre_2
0          Action    Adventure
1          Action    Adventure
2          Action    Adventure
3          Action     Thriller
5          Action    Adventure
...           ...          ...
5033        Drama       Sci-Fi
5034     Thriller     Thriller
5035       Action        Crime
5037       Comedy        Drama
5042  Documentary  Documentary

[3856 rows x 2 columns]

In [ ]:
# Add genres columns to movies dataframe
df_movies=pd.concat([df_movies,Temp_genre],axis=1)
df_movies

director_name  num_critic_for_reviews       gross  \
0         James Cameron                   723.0  760.505847   
1        Gore Verbinski                   302.0  309.404152   
2            Sam Mendes                   602.0  200.074175   
3     Christopher Nolan                   813.0  448.130642   
5        Andrew Stanton                   462.0   73.058679   
...                 ...                     ...         ...   
5033      Shane Carruth                   143.0    0.424760   
5034   Neill Dela Llana                    35.0    0.070071   
5035   Robert Rodriguez                    56.0    2.040920   
5037       Edward Burns                    14.0    0.004584   
5042           Jon Gunn                    43.0    0.085222   

                                   genres     actor_1_name  \
0         Action|Adventure|Fantasy|Sci-Fi      CCH Pounder   
1                Action|Adventure|Fantasy      Johnny Depp   
2               Action|Adventure|Thriller  Christoph Waltz   
3                         Action|Thriller        Tom Hardy   
5                 Action|Adventure|Sci-Fi     Daryl Sabara   
...                                   ...              ...   
5033                Drama|Sci-Fi|Thriller    Shane Carruth   
5034                             Thriller      Ian Gamazon   
5035  Action|Crime|Drama|Romance|Thriller  Carlos Gallardo   
5037                         Comedy|Drama      Kerry Bishé   
5042                          Documentary      John August   

                                    movie_title  num_voted_users  \
0                                       Avatar            886204   
1     Pirates of the Caribbean: At World's End            471220   
2                                      Spectre            275868   
3                        The Dark Knight Rises           1144337   
5                                  John Carter            212204   
...                                         ...              ...   
5033                                    Primer             72639   
5034                                    Cavite               589   
5035                               El Mariachi             52055   
5037                                 Newlyweds              1338   
5042                         My Date with Drew              4285   

      num_user_for_reviews language    budget  title_year  imdb_score  \
0                   3054.0  English  237.0000      2009.0         7.9   
1                   1238.0  English  300.0000      2007.0         7.1   
2                    994.0  English  245.0000      2015.0         6.8   
3                   2701.0  English  250.0000      2012.0         8.5   
5                    738.0  English  263.7000      2012.0         6.6   
...                    ...      ...       ...         ...         ...   
5033                 371.0  English    0.0070      2004.0         7.0   
5034                  35.0  English    0.0070      2005.0         6.3   
5035                 130.0  Spanish    0.0070      1992.0         6.9   
5037                  14.0  English    0.0090      2011.0         6.4   
5042                  84.0  English    0.0011      2004.0         6.6   

      movie_facebook_likes      profit      genre_1      genre_2  
0                    33000  523.505847       Action    Adventure  
1                        0    9.404152       Action    Adventure  
2                    85000  -44.925825       Action    Adventure  
3                   164000  198.130642       Action     Thriller  
5                    24000 -190.641321       Action    Adventure  
...                    ...         ...          ...          ...  
5033                 19000    0.417760        Drama       Sci-Fi  
5034                    74    0.063071     Thriller     Thriller  
5035                     0    2.033920       Action        Crime  
5037                   413   -0.004416       Comedy        Drama  
5042                   456    0.084122  Documentary  Documentary  

[3856 row

**Group the dataframe using genres_1 as the primary column and genres_2 as the secondary column**

In [ ]:
df_movies.groupby(['genre_1','genre_2']).gross.mean().sort_values(ascending=False).head(5)

genre_1    genre_2  
Family     Sci-Fi       434.949459
Adventure  Sci-Fi       228.627758
           Family       118.919540
           Animation    116.998550
Action     Adventure    109.595465
Name: gross, dtype: float64

**Observations:** Here family+ sci-Fi is most popular combo genres out there means it collect more amount.

### **Find the critic-favourite and audience-favourite actors**

In [ ]:
df_movies['actor_1_name'].value_counts()

Robert De Niro               42
Johnny Depp                  38
J.K. Simmons                 31
Denzel Washington            30
Nicolas Cage                 30
                             ..
Martin Dew                    1
Chriss Anglin                 1
Kevin Alejandro               1
Catherine Lough Haggquist     1
John August                   1
Name: actor_1_name, Length: 1509, dtype: int64

In [ ]:
Meryl_Streep=df_movies[df_movies['actor_1_name']=='Meryl Streep']
Leonardo_DiCaprio=df_movies[df_movies['actor_1_name']=='Leonardo DiCaprio']
Brad_Pitt=df_movies[df_movies['actor_1_name']=='Brad Pitt']

In [ ]:
combined=Brad_Pitt.append([Meryl_Streep,Leonardo_DiCaprio])
combined

<ipython-input-55-52fd1d4a30a5>:1: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  combined=Brad_Pitt.append([Meryl_Streep,Leonardo_DiCaprio])


director_name  num_critic_for_reviews       gross  \
101           David Fincher                   362.0  127.490802   
147       Wolfgang Petersen                   220.0  133.228348   
254       Steven Soderbergh                   198.0  125.531634   
255              Doug Liman                   233.0  186.336103   
382              Tony Scott                   142.0    0.026871   
400       Steven Soderbergh                   186.0  183.405771   
470              David Ayer                   406.0   85.707116   
611     Jean-Jacques Annaud                    76.0   37.901509   
683           David Fincher                   315.0   37.023395   
792         Patrick Gilmore                    98.0   26.288320   
940             Neil Jordan                   120.0  105.264608   
1490        Terrence Malick                   584.0   13.303319   
1722         Andrew Dominik                   273.0    3.904982   
2204  Alejandro G. Iñárritu                   285.0   34.300771   
2333    Angelina Jolie Pitt                   131.0    0.531009   
2682         Andrew Dominik                   414.0   14.938570   
2898             Tony Scott                   122.0   12.281500   
410            Nancy Meyers                   187.0  112.703470   
1106          Curtis Hanson                    42.0   46.815748   
1204            Nora Ephron                   252.0   94.125426   
1408          David Frankel                   208.0  124.732962   
1483         Robert Redford                   227.0   14.998070   
1575         Sydney Pollack                    66.0   87.100000   
1618          David Frankel                   234.0   63.536011   
1674          Carl Franklin                    64.0   23.209440   
1925         Stephen Daldry                   174.0   41.597830   
2781         Phyllida Lloyd                   331.0   29.959436   
3135          Robert Altman                   211.0   20.338609   
26            James Cameron                   315.0  658.672302   
50             Baz Luhrmann                   490.0  144.812796   
97        Christopher Nolan                   642.0  292.568851   
179   Alejandro G. Iñárritu                   556.0  183.635922   
257         Martin Scorsese                   267.0  102.608827   
296       Quentin Tarantino                   765.0  162.804648   
307            Edward Zwick                   166.0   57.366262   
308         Martin Scorsese                   606.0  116.866727   
326         Martin Scorsese                   233.0   77.679638   
361         Martin Scorsese                   352.0  132.373442   
452         Martin Scorsese                   490.0  127.968405   
641            Ridley Scott                   238.0   39.380442   
911        Steven Spielberg                   194.0  164.435221   
990             Danny Boyle                   118.0   39.778599   
1114             Sam Mendes                   323.0   22.877808   
1422        Randall Wallace                    83.0   56.876365   
1453         Clint Eastwood                   392.0   37.304950   
1560              Sam Raimi                    63.0   18.636537   
2067             Jerry Zaks                    45.0   12.782508   
2757           Baz Luhrmann                   106.0   46.338728   
3476           Baz Luhrmann                   490.0  144.812796   

                                                 genres       actor_1_name  \
101                               Drama|Fantasy|Romance          Brad Pitt   
147                                           Adventure          Brad Pitt   
254                                      Crime|Thriller          Brad Pitt   
255                Action|Comedy|Crime|Romance|Thriller          Brad Pitt   
382                               Action|Crime|Thriller          Brad Pitt   
400                                      Crime|Thriller          Brad Pitt   
470                                    Action|Drama|War          Brad Pitt   
611               Adventure|Biography|D

In [ ]:
combined.groupby('actor_1_name').num_critic_for_reviews.mean()

actor_1_name
Brad Pitt            245.000000
Leonardo DiCaprio    330.190476
Meryl Streep         181.454545
Name: num_critic_for_reviews, dtype: float64

In [ ]:
combined.groupby('actor_1_name').num_user_for_reviews.mean()

actor_1_name
Brad Pitt            742.352941
Leonardo DiCaprio    914.476190
Meryl Streep         297.181818
Name: num_user_for_reviews, dtype: float64

In [ ]:
# Alternate way to find mean for both columns
combined.groupby('actor_1_name')[['num_user_for_reviews','num_critic_for_reviews']].mean()

num_user_for_reviews  num_critic_for_reviews
actor_1_name                                                   
Brad Pitt                    742.352941              245.000000
Leonardo DiCaprio            914.476190              330.190476
Meryl Streep                 297.181818              181.454545

**Observations:** Above from two lists Leonardo DiCapriogot high reviews.

https://www.youtube.com/watch?v=lFGtBC2MG8Y  -Reference link